### ISSUE TITLE
Feature request: export `Records` schema building blog for customizations

### ISSUE THREAD

**Issue by am29d** [aws-powertools/powertools-lambda-typescript#2811]
### Use case

When you want to extend the built-in schema with your custom schema you can extend the it and overwrite specific parts with of the built-in schema we provide. 

```typescript
const mySchema = z.object({
  name: z.string(),
  age: z.number()
});

const AlbCustomizedSchema = AlbSchema.extend({
  body: mySchema
});
```

The `body` value is `z.string()` in the `AlbSchema` and we can plug in modifications. 

This works for schemas with first the modifiable attribute available in the main schema, such as `AlbSchema`, `EventBrdigeSchema` or others. 

When working with schemas that contain a list of elements the `body` part is not accessible, thus customers can't extend it without replicating the entire `Record`. For example, the `SqsEvent` is:

```json
  {
    "Records": [
      {
        "messageId": "059f36b4-87a3-44ab-83d2-661975830a7d",
        "receiptHandle": "AQEBwJnKyrHigUMZj6rYigCgxlaS3SLy0a...",
        "body": "Test message.",
        "attributes": {
          "ApproximateReceiveCount": "1",
          "SentTimestamp": "1545082649183",
          "SenderId": "AIDAIENQZJOLO23YVJ4VO",
          "ApproximateFirstReceiveTimestamp": "1545082649185"
        },
        "messageAttributes": {
          "testAttr": {
            "stringValue": "100",
            "binaryValue": "base64Str",
            "dataType": "Number"
          }
        },
        "md5OfBody": "e4e68fb7bd0e697a0ae8f1bb342846b3",
        "eventSource": "aws:sqs",
        "eventSourceARN": "arn:aws:sqs:us-east-2:123456789012:my-queue",
        "awsRegion": "us-east-2"
      }
    ]
  }
  ```

To overwrite the body, we'd need:

```typescript
const SqsCustomizedSchema = SqsSchema.extend({
 Records: z.array(
     z.object({
         messageId...
         ...
         body: mySchema,
         // the rest
     })
  )
})
```

### Solution/User Experience

Export the record elements that are nested parts of the built-in schema. For SQS it would be https://github.com/aws-powertools/powertools-lambda-typescript/blob/1e7175a8ff201b821bc72a8725994b34e653244f/packages/parser/src/schemas/sqs.ts#L26-L38

This would allow us change only the body field

```typescript
const SqsMySchema = SqsSchema.extend({
  Records: SqsRecordSchema.extend({
   body: mySchema
  })
})
```

The requirement for nested `Records` exports applies to:

* [KafkaRecordSchema](https://github.com/aws-powertools/powertools-lambda-typescript/blob/1e7175a8ff201b821bc72a8725994b34e653244f/packages/parser/src/schemas/kafka.ts#L3)
* [KinesisFirehoseRecrod](https://github.com/aws-powertools/powertools-lambda-typescript/blob/1e7175a8ff201b821bc72a8725994b34e653244f/packages/parser/src/schemas/kinesis-firehose.ts#L25)
* [KinesisFirehoseSqsSchema](https://github.com/aws-powertools/powertools-lambda-typescript/blob/1e7175a8ff201b821bc72a8725994b34e653244f/packages/parser/src/schemas/kinesis-firehose.ts#L31)
* [KinesisDataStreamRecord](https://github.com/aws-powertools/powertools-lambda-typescript/blob/1e7175a8ff201b821bc72a8725994b34e653244f/packages/parser/src/schemas/kinesis.ts#L29)
* [SesRecordSchema](https://github.com/aws-powertools/powertools-lambda-typescript/blob/1e7175a8ff201b821bc72a8725994b34e653244f/packages/parser/src/schemas/ses.ts#L55)
* [SnsRecordSchema](https://github.com/aws-powertools/powertools-lambda-typescript/blob/1e7175a8ff201b821bc72a8725994b34e653244f/packages/parser/src/schemas/sns.ts#L61) and [SnsNotificationSchema](https://github.com/aws-powertools/powertools-lambda-typescript/blob/1e7175a8ff201b821bc72a8725994b34e653244f/packages/parser/src/schemas/sns.ts#L8)
* [SqsRecordSchema](https://github.com/aws-powertools/powertools-lambda-typescript/blob/1e7175a8ff201b821bc72a8725994b34e653244f/packages/parser/src/schemas/sqs.ts#L26)

While some of this objects are exported in their own files, they are not part of the barrel export and thus customers can't import them. 

### Acknowledgment

- [X] This feature request meets [Powertools for AWS Lambda (TypeScript) Tenets](https://docs.powertools.aws.dev/lambda/typescript/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Python](https://github.com/aws-powertools/powertools-lambda-python/), [Java](https://github.com/aws-powertools/powertools-lambda-java/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### Future readers

Please react with 👍 and your use case to help us understand customer demand.

**Comment by dreamorosi** [comment]
Hi @am29d, this is a good point and we should definitely expose these schemas.

My only concern is with where to export them. On one hand I see the value of exposing them in the main barrel file both for consistency and relative ease of usage. On the other hand, I think we should help customers do the right thing and instead provide scoped exports.

Even when using ESM with tree-shaking enabled, using the barrel file causes all the schemas to be pulled in into the bundle. This is because `esbuild` cannot know which one will actually be used at runtime, it only "sees" that you've required the entire barrel file. The same is true for envelopes.

Exporting schemas in their own subpath export like we [started doing with API Gateway-related schemas](https://github.com/aws-powertools/powertools-lambda-typescript/blob/8b0cae45d007e1edb50922e1beabd8384c9d1687/packages/parser/package.json#L46-L49) would instead allow customers to target a specific group of schemas and bring in only these.

I don't have concrete data to back this up, but I would imagine that customers won't be using more than a handful built-in schemas in each function. If this is true, then giving scoped exports will help to keep their bundle size small, especially as we add more and more specialized schemas.

So in short, I am in favor of exporting these schemas as they improve the DX, but I would suggest we do this responsibly and provide subpath exports for each service.

**Comment by am29d** [comment]
I agree, I have seen your changes for API gateway and we should change other exports as well, i.e.:


**before**
```typescript
import { SqsSchema } from '@aws-lambda-powertools/parser/schemas';
```

**after**
```typescript
import { SqsSchema } from '@aws-lambda-powertools/parser/schemas/sqs';
```

**Comment by dreamorosi** [comment]
As suggested in #2829 by @am29d, this issue & that one are very much related.

I think it'd be useful to complete this before merging that one, so we can include the new schemas in the subpath exports we'll add in #2829.

**Comment by daschaa** [comment]
@am29d @dreamorosi Should i take this issue first then? :)

**Comment by am29d** [comment]
> @am29d @dreamorosi Should i take this issue first then? :)

@daschaa of course, all yours :)
